indra/newview/lluilistener.cpp

```diff
@@ -59,43 +60,57 @@ LLUIListener::LLUIListener():
 
 void LLUIListener::call(const LLSD& event) const
 {
-    LLUICtrl::commit_callback_t* func =
-        LLUICtrl::CommitCallbackRegistry::getValue(event["function"]);
-    if (! func)
+    static F64 last_throttle_time = 0.0;
+    Response response(LLSD(), event);
+    LLUICtrl::LLCommitCallbackInfo *info = LLUICtrl::SharedCommitCallbackRegistry::getValue(event["function"]);
+    if (!info )
     {
-        // This API is intended for use by a script. It's a fire-and-forget
-        // API: we provide no reply. Therefore, a typo in the script will
-        // provide no feedback whatsoever to that script. To rub the coder's
-        // nose in such an error, crump rather than quietly ignoring it.
-        LL_WARNS("LLUIListener") << "function '" << event["function"] << "' not found" << LL_ENDL;
+        response.error(STRINGIZE("Function " << std::quoted(event["function"].asString()) << "was not found"));
+        return;
     }
-    else
+    if (info->mHandleUntrusted == LLUICtrl::LLCommitCallbackInfo::UNTRUSTED_BLOCK) 
+    {
+        response.error(STRINGIZE("Function " << std::quoted(event["function"].asString()) << " is not allowed to be called from the script"));
+        return;
+    }
+    if (info->mHandleUntrusted == LLUICtrl::LLCommitCallbackInfo::UNTRUSTED_THROTTLE)
+    {
+        F64 cur_time = LLTimer::getElapsedSeconds();
+        if (cur_time < last_throttle_time + THROTTLE_PERIOD)
+        {
+            LL_WARNS("LLUIListener") << "Throttled function " << std::quoted(event["function"].asString()) << LL_ENDL;
+            return;
+        }
+        last_throttle_time = cur_time;
+    }
```

Comment:
Are we going to need to replicate support for `UNTRUSTED_BLOCK` and `UNTRUSTED_THROTTLE` in other places, or is this the one place where that's applicable?

There's another instance of new throttle code in a recent PR. I find myself musing about a wrapper like:
```
// either execute 'lambda' and return 'true', or don't and return 'false'
// shared throttle
template <typename LAMBDA>
bool throttle(F64& lasttime, F64 limit, LAMBDA&& lambda);
// function-static throttle (distinct instance for each LAMBDA)
template <typename LAMBDA>
bool throttle(F64 limit, LAMBDA&& lambda)
{
    static F64 lasttime = 0.0;
    return throttle(lasttime, limit, std::forward<LAMBDA>(lambda));
}
```